### PR TITLE
WIP/RFC: introduce global clock, use in lease manager

### PR DESCRIPTION
**Pull request by axw**
This PR introduces a global clock document to the database. We run (no more than) one worker in the controller cluster to update the global clock. We guarantee that the global clock is monotonic, and also that a global clock second is *at least* as long as a wall clock second. If consumers take this into account, then they should not be impacted if there is a period in which the global clock is not updated.

We pass a global clock reader into the state/lease client, so that it can compute expiry times based on the global time. Lease documents store the start time as a global time, and the request duration; we then make the expiry time relative to the local time, for use by clients. The local time is expected to contain a monotonic component (i.e. Go 1.9+), so it can be compared to time.Now().

We refresh the global clock time when a claim or extension is made, to guarantee that the lease is held for at least the specified duration. When a lease is to be expired, we first attempt to use the most recent global clock time, and then refresh if the lease expires after the cached time, only failing if the lease expires after the refreshed time.

The global clock updater currently updates every second. We can increase the period to reduce database pressure and replication overhead, at a cost of less timely reaction to lease expiry.

### CODE REVIEW

File: state/globalclock/schema.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2015 Canonical Ltd.
```
Comment (wallyworld) [inline state/globalclock/schema.go]: 2017

Comment (axw) [reply state/globalclock/schema.go]: fixed

**Comment by axw** [comment]
Closing this, but leaving for history. I'll create a new PR with final code.

**Comment by axw** [comment]
https://github.com/juju/juju/pull/7957

**Pull request closed without merging**